pkg/workloads/cortex/lib/api/batching.py

```diff
@@ -0,0 +1,129 @@
+# Copyright 2020 Cortex Labs, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import threading as td
+import time
+import traceback
+from collections import defaultdict
+from http import HTTPStatus
+from typing import Any, Callable, Dict, List
+
+from starlette.responses import Response
+
+from ..exceptions import UserRuntimeException
+from ..log import cx_logger as logger
+
+
+class DynamicBatcher:
+    def __init__(self, predictor_impl: Callable, max_batch_size: int, batch_interval: int):
+        self.predictor_impl = predictor_impl
+
+        self.batch_max_size = max_batch_size
+        self.batch_interval = batch_interval  # measured in seconds
+
+        # waiter synchronizes threads and prevents that extra threads
+        # modify the batch while prediction is in process
+        self.waiter = td.Event()
+        self.waiter.set()
+
+        self.barrier = td.Barrier(self.batch_max_size + 1)
+
+        self.samples = {}
+        self.predictions = {}
+        td.Thread(target=self._batch_engine).start()
+
+    def _batch_engine(self):
+        while True:
+            if len(self.predictions) > 0:
```

Comment:
One of the users using batch has complained about performance issues when using the serverside batching implementation. It may be possible that the busy waiting is adding unnecessary load to the CPU. It may good idea to make a ticket for this to followup on.